### ISSUE TITLE
[Suggestion] Reverse Connection Request

### ISSUE THREAD

**Issue by jericomanapsal** [ipfs/ipfs#336]
Hello

**All IPv4**

I have two nodes on my LAN under NAT. I want both of my nodes to be able to connect to the swarm and be connected on. I can port forward 4001 to my first node, and use 4002 or other ports to the second node. However, only the first node is able to be connected upon. Sure the files on the second node can traverse through the first node, but see the next case.

A relatively big house network consists of four subnets but I'm the only one operating a node under one subnet. I have no control over the network configuration on this house. Whenever needed, the router forwards a port to my node. But the port selection depends on the availability, usually higher than 21000. Say it assigns 21467. This means that my node is reachable through the address:
```
ipfs swarm connect /ipv4/public_ip/tcp/21467/my_hash
```
While this may work, most if not all nodes that currently operates on the swarm will use 4001 which may render my node unreachable.

Is there a way to request to at least one peer to connect to my node through the port that I will give? Something like:
```
ipfs swarm reverse other_peer_id /ipv4/public_ip/tcp/21467/my_hash
```
I just need at least one node (my hosted node) that automatically pins whatever new hashes I have, to connect to my node.

**Comment by Stebalien** [comment]
> While this may work, most if not all nodes that currently operates on the swarm will use 4001 which may render my node unreachable.

That's just the default. You should be able to run your node on any port. If you change the port, your node will announce a different address and other nodes will connect to that different address.

**Comment by jericomanapsal** [comment]
Will my node broadcast my public IP together with the detected port provided by the router?

**Comment by Stebalien** [comment]
Yes. Whenever you connect to another node, that node will tell you the address it observes. You'll then announce these "observed" addresses to the network.

go-ipfs also tries to reuse the source port to maintain the port mapping.
